### DISCUSSION TITLE
PLEASE! … create a how-to-tutorial on Chaquopy's/briefcases' use of external binaries in Android apps

### DISCUSSION THREAD

**Discussion by clemsam** [beeware/briefcase#958] [Q&A]
On my research how to get sound into an (existing) briefcase-Android project's app ([at Stack-O](https://stackoverflow.com/questions/74120614/how-to-play-sound-short-sentence-in-an-android-app-created-by-beeware-using-py/74275351?noredirect=1#comment131175935_74275351)) I got the recommendation from BeeWare's new apiarist Malcolm Smith (creator of the amazing Chaquopy Python APIs) to use his API and SoundPool for my project.
In [Tutorial 6](https://docs.beeware.org/en/latest/tutorial/tutorial-6.html) there is stated quite definitely _"you can only use pure Python packages"_ and _"This means that libraries like **numpy, scikit-learn, or cryptography** can be used in a desktop app, but not a mobile app."_

But fortunately [briefcases FAQ](https://briefcase.readthedocs.io/en/latest/background/faq.html) mentions the [**_"Chaquopy package index"_**](https://chaquo.com/pypi-7.0/) website where you can find a list of nearly a hundred packages – among them:
[**_numpy, scikit-learn_**](), and [**_cryptography_**]() …

As a non-professional hobbyist and Python-enthusiast I'd really, really appreciate it **_very-very-very much_** if someone could develop and post a simple example/tutorial here at beeware on how to go about including & implementing e.g. **_matplotlib_** into an Android app, covering items like:
– pip --… install matplotlib from Chaquopy's repository (binary wheels)
– editing build.gradle, requirements.txt (and whereever else necessary)
– importing matplotlib into app.py and a simple use of it
– whatever else is relevant to an app
(– maybe also: how to detect which version is required ?)

_Thanks a lot!  Clemsam_

**Comment by mhsmith** [comment]
Yes, the tutorial page is out of date: you can now use all the Chaquopy binary packages on Android.

If your app uses Briefcase, just add the package to the `requires` section of your pyproject.toml file, as shown in [the BeeWare tutorial](https://docs.beeware.org/en/latest/tutorial/tutorial-6.html#updating-dependencies), and Briefcase will deal with all the details of pip and build.gradle. You'll need to be using Briefcase 0.3.10 or later, and preferably run `briefcase create` with Python 3.8. As you can see from the package index, Python 3.9 and 3.10 have a smaller selection of packages, while Python 3.11 has almost nothing yet.

Or if your app uses Chaquopy without Briefcase, follow the instructions in the Chaquopy documentation [here](https://chaquo.com/chaquopy/doc/current/android.html#android-requirements).

**Reply by aryan1165** [reply]
Hi @mhsmith, I have build wheels for Ctranslate2 and have raised a PR in chaquopy's repository, but it is not merged, How can i use the local copy of chaquopy repository to integrate with briefcase and integrate with android mobile devices.

**Reply by mhsmith** [reply]
See #1270. At least for Android, the "URL" can also be a local path.

**Reply by kfchou** [reply]
Related to this question, I have 2 wheels I'd like to use:

In briefcase' pyproject.toml file:
```
[tool.briefcase.app.numpy_test.android]
requires = [
    "toga-android~=0.5.0",
    "file:///home/kfchou/repos/my_numpy/numpy/android_test/numpy_test/wheels/numpy-2.4.0.dev0-cp313-cp313-android_21_arm64_v8a.whl",
    "file:///home/kfchou/repos/my_numpy/numpy/android_test/numpy_test/wheels/numpy-2.4.0.dev0-cp313-cp313-android_21_x86_64.whl",
]

pip_install_args = [
    "--extra-index-url=/home/kfchou/repos/my_numpy/numpy/android_test/numpy_test/wheels/",
    ]
```
Note that the android_21 tag is incorrect; I had specified android_30. This is an issue on my end. 

Nevertheless, the build fails with:
```
> Task :app:generateDebugPythonRequirements
Chaquopy: Installing for arm64-v8a
Looking in indexes: https://pypi.org/simple, https://chaquo.com/pypi-13.1
Collecting toga-android~=0.5.0
  Using cached toga_android-0.5.2-py3-none-any.whl (69 kB)
Processing /home/kfchou/repos/my_numpy/numpy/android_test/numpy_test/wheels/numpy-2.4.0.dev0-cp313-cp313-android_21_arm64_v8a.whl
Collecting toga-core==0.5.2
  Using cached toga_core-0.5.2-py3-none-any.whl (145 kB)
Collecting travertino==0.5.2
  Using cached travertino-0.5.2-py3-none-any.whl (24 kB)
Installing collected packages: travertino, toga-core, toga-android, numpy
Successfully installed numpy-2.4.0.dev0 toga-android-0.5.2 toga-core-0.5.2 travertino-0.5.2
Chaquopy: Installing for x86_64
Looking in indexes: https://pypi.org/simple, https://chaquo.com/pypi-13.1
ERROR: Could not find a version that satisfies the requirement numpy==2.4.0.dev0 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23.4, 1.23.5, 1.24.0, 1.24.1, 1.24.2, 1.24.3, 1.24.4, 1.25.0, 1.25.1, 1.25.2, 1.26.2, 1.26.3, 1.26.4, 2.0.0, 2.0.1, 2.0.2, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, 2.3.0, 2.3.1, 2.3.2)
ERROR: No matching distribution found for numpy==2.4.0.dev0
Chaquopy: Exit status 1

> Task :app:generateDebugPythonRequirements FAILED
```
How come the build process is not honoring the x86_64.whl file?

I haven't tried to build numpy 2.3.2. Would that be the issue?

It is possible that the wheels are not compiled correctly; I'm trying to test them. The best way I thought to do this was to build a briefcase example project (`briefcase start`) and ask it to import numpy.

Please advise

**Reply by mhsmith** [reply]
You can't mix wheels for multiple architectures in your `requires` list, because each architecture is installed with a separate run of `pip`. Instead, your `requires` list should simply say `numpy==2.4.0.dev0`.

There is no `pip_install_args` setting. Where did you get that from? The correct setting is [`requirement_installer_args`](https://briefcase.readthedocs.io/en/stable/reference/configuration.html#requirement_installer_args).

Finally, for pip's `--extra-index-url` option to work, you'll need to put the wheels inside a `numpy` subdirectory. Alternatively, you can use `--find-links`, which expects a flat directory layout. See the `requirement_installer_args` documentation for examples.

**Reply by kfchou** [reply]
Thanks for your response @mhsmith . I've updated my configurations but the build is still failing. My `pyproject.toml` file has
```
[tool.briefcase.app.numpy_test.android]
requires = [
    "toga-android~=0.5.0",
    "numpy==2.4.0.dev0",
]

requirement_install_args = [
    "--find-links", "./wheels/numpy",
    ]

# also tried

requirement_install_args = [
    "--extra-index-url", "./wheels",
    ]

```

and my directory structure is
```
.
├── CHANGELOG
├── LICENSE
├── README.rst
├── build
│   └── numpy_test
│       └── android
├── logs
├── pyproject.toml <----- config file
├── src
│   └── numpy_test
│       ├── __init__.py
│       ├── __main__.py
│       ├── app.py
│       └── resources
├── tests
│   ├── __init__.py
│   ├── numpy_test.py
│   └── test_app.py
└── wheels
    └── numpy
        ├── numpy-2.4.0.dev0-cp313-cp313-android_21_arm64_v8a.whl
        └── numpy-2.4.0.dev0-cp313-cp313-android_21_x86_64.whl
```

The build log suggests pip is still looking in pypi:
```
> Task :app:generateDebugPythonRequirements
Chaquopy: Installing for arm64-v8a
Looking in indexes: https://pypi.org/simple, https://chaquo.com/pypi-13.1
Collecting toga-android~=0.5.0
  Using cached toga_android-0.5.2-py3-none-any.whl (69 kB)
ERROR: Could not find a version that satisfies the requirement numpy==2.4.0.dev0 (from -r requirements.txt (line 3)) (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23.4, 1.23.5, 1.24.0, 1.24.1, 1.24.2, 1.24.3, 1.24.4, 1.25.0, 1.25.1, 1.25.2, 1.26.2, 1.26.3, 1.26.4, 2.0.0, 2.0.1, 2.0.2, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, 2.3.0, 2.3.1, 2.3.2)
ERROR: No matching distribution found for numpy==2.4.0.dev0 (from -r requirements.txt (line 3))
Chaquopy: Exit status 1
```

Is there something I'm missing?

**Reply by mhsmith** [reply]
This is a separate question to the original discussion, so I've moved it to #2472.
